app/controllers/reports/external_referrals_controller.rb

```diff
@@ -34,12 +34,26 @@ def refer
 
     private def index_response
       respond_to do |format|
-        format.html {}
+        format.html do
+          # paginate, but attempt to get all clients on one page because we may be adding a referral,
+          # which won't work right if we're exporting a huge number all at once.
+          @clients = @clients.page(params[:page]).per(250)
+        end
         format.xlsx do
           @clients = @clients.preload(:external_referrals, :active_matches, :project_client).to_a
-          @dv = @clients.select { |c| c.domestic_violence == 1 }
-          @sheltered = @clients.select(&:majority_sheltered)
-          @unsheltered = @clients.reject(&:majority_sheltered)
+          @dv = @clients.select { |c| c.domestic_violence == 1 }.map(&:id).to_set
+          @sheltered = @clients.select(&:majority_sheltered).map(&:id).to_set
+          @unsheltered = @clients.reject(&:majority_sheltered).map(&:id).to_set
+          # Get a hash of assessments that we can use for generating columns, we only need one per assessment type, ignore the differences between
+          # de-identified and identified
+          @assessment_classes = {}
+          @clients.each do |client|
+            assessment = NonHmisAssessment.to_class(client.assessment_name)
+            title = NonHmisAssessment.title_from_type_for_matching(client.assessment_name).gsub(' - Identified', '').gsub(' - Deidentified', '')
```

Comment:
I'd like to see the `gsubs` pushed down a level, maybe `def parent_assessment_name`? Rather than have the details about how the assessments get labeled end up in here.